src/merkle/merkle_tree.rs

```diff
@@ -0,0 +1,278 @@
+        Self::create_tree(transactions)
+    }
+
+    fn create_parent_from_siblings(
+        left_son: Option<Box<MerkleNode>>,
+        mut right_son: Option<Box<MerkleNode>>,
+    ) -> MerkleNode {
+        let mut hasher = DefaultHasher::new();
+
+        if let Some(left_sibling) = &left_son {
+            left_sibling.hash_value.hash(&mut hasher);
+            if let Some(right_sibling) = &right_son {
+                right_sibling.hash_value.hash(&mut hasher);
+            } else {
+                right_son = left_son.clone();
+                left_sibling.hash_value.hash(&mut hasher);
+            }
+        }
+        MerkleNode::new(hasher.finish(), left_son, right_son)
+    }
+
+    fn create_tree(transactions: Vec<H>) -> MerkleTree<H> {
+        if transactions.is_empty() {
+            return Self {
+                merkle_root: MerkleNode::new(0, None, None),
+                leafs: transactions,
+            };
+        }
+
+        let transactions_hash = Self::get_hashes_of_transactions(&transactions);
+
+        let mut nodes = Vec::new();
+        for hash in transactions_hash {
+            nodes.push(Box::new(MerkleNode::new(hash, None, None)));
+        }
+
+        while nodes.len() > 1 {
+            let mut parents = Vec::new();
+
+            for _ in (0..nodes.len()).step_by(2) {
+                let parent = Self::create_parent_from_siblings(nodes.pop(), nodes.pop());
+                parents.push(Box::new(parent));
+            }
+
+            nodes = parents;
+        }
+
+        Self {
+            merkle_root: *nodes[0].clone(),
+            leafs: transactions,
+        }
+    }
+
+    fn get_hashes_of_transactions(transactions: &Vec<H>) -> Vec<u64> {
+        let mut transactions_hash = Vec::new();
+        for transaction in transactions {
+            let mut hasher = DefaultHasher::new();
+            transaction.hash(&mut hasher);
+            let transaction_hash = hasher.finish();
+            transactions_hash.push(transaction_hash);
+        }
+        transactions_hash
+    }
+
+    pub fn verify(&mut self, transaction: H, proof: Vec<SiblingsHash>) -> bool {
+        let mut hasher = DefaultHasher::new();
+        transaction.hash(&mut hasher);
+        let mut transaction = hasher.finish();
+        for sibling_hash in proof {
+            hasher = DefaultHasher::new();
+            match sibling_hash {
+                SiblingsHash::LeftSibling(left_hash) => {
+                    left_hash.hash(&mut hasher);
+                    transaction.hash(&mut hasher);
+                }
+                SiblingsHash::RightSibling(right_hash) => {
+                    transaction.hash(&mut hasher);
+                    right_hash.hash(&mut hasher);
+                }
+            }
+
+            transaction = hasher.finish();
+        }
+
+        transaction == self.merkle_root.hash_value
+    }
+
+    fn recursive_get_proof(
+        actual_node: &MerkleNode,
```

Comment:
I believe it should be "current", not "actual". "Actual" in english is not "actual" in spanish.

Reply:
Oh really? Didn't know about that one, I'm changing that